### DISCUSSION TITLE
Integrate Godot + GodotFirebase with React app.

### DISCUSSION THREAD

**Discussion by smyrgeorge** [GodotNuts/GodotFirebase#193] [Q&A] [answered]
Hello!

I have created an application using `React + Firebase`.

- I'm authenticating the users (in the react app) using
`.signInWithPopup(new firebase.auth.GoogleAuthProvider())`.
- After a successful authentication, i obtain a json object that contains the following:
```
{
	"additionalUserInfo": {
		"providerId": "google.com",
		"isNewUser": false,
		"profile": {}
	},
	"credential": {
		"a": null,
		"accessToken": "ya29.a0.....",
		"idToken": "eyJhbGci.....",
		"providerId": "google.com",
		"signInMethod": "google.com"
	},
	"user": {}
}
```

Can i use the `idToken` (or `accessToken`) to authenticate to firebase inside the Godot.
So, i want to obtain the tokens inside the Godot using the `Javascript.eval` function.
I tried `Firebase.Auth.login_with_oauth(token)` but none of these tokens worked.

Any ideas?

PS: GodotFirebase  seems to be very powerful 😄 💪

**Comment by BearDooks** [comment, marked as answer]
@smyrgeorge you might be able to use the func `login_with_custom_token`. This is a new function that I think was just put in with one of the latest PRs. 

I would try that and let us know what happens. 

Chuck

**Reply by smyrgeorge** [reply]
@BearDooks, thanks for your quick reply!

`login_with_custom_token` does not seem to work in this case, an error occurred.

```
error code: 400
message: MISSING_IDENTIFIER
```

This could may work in a total different flow (without using the google OAuth).
For example, you could have a third party app (eg. a lambda function) that creates custom tokens and then use the tokens in both React app and Godot. But in this scenario it's difficult to use third party authentication providers (such  as Facebook, Apple, etc.).

**Reply by BearDooks** [reply]
Sorry it took me so long to get back to you. 

This is some good info. Let me go back to the team and see what they think.

Chuck

**Reply by BearDooks** [reply]
@smyrgeorge I spoke with the team and have some new info that might help.

The MISSING_IDENTIFIER error is related to how the token was created. It means that the payload to create the token is missing something, like the user uid. https://firebase.google.com/docs/auth/admin/create-custom-tokens#create_custom_tokens_using_a_third-party_jwt_library

The user mars3142 has been able to use a custom token, and they might be able to shed some more light on it. While we did test it, we don't currently have a project that uses it. (something we should probably through into our backlog) That user is talking about it here: https://github.com/GodotNuts/GodotFirebase/discussions/189


There is another function for logging in with a custom token, but I think currently it only works with Google.
